### ISSUE TITLE
Predictions on custom object. Position looks ok, but visualization does not work

### ISSUE THREAD

**Issue by apirrone** [megapose6d/megapose6d#12]
Hello, 

First, everything works just fine on the barbecue-sauce example provided.

I am trying to use megapose6d on my own objects. Here is my process : 

- I take multiple photos of the object and reconstruct a 3d mesh using MeshRoom
- I then crop the reconstructed scene to keep only the object of interest (using CloudCompare or MeshLab) 
- I resize the mesh so that the dimensions correspond to the real size of the object millimeters 
- I export the object mesh in ply

I created an example directory like your barbecue-sauce, provided `camera_data.json`, `image_rgb.png`, an `input/object_data.json` file with the 2d bounding box and the `mesh/myobject/myobject.ply;texture.png` files. 

When I run the inference script, I get the following `output/object_data.json`:

```json
[
    {
        "label": "myobject",
        "TWO": [
            [
                0.0793891647744044,
                -0.5925398067911702,
                -0.12802647375990645,
                0.7913300772232438
            ],
            [
                0.03524733707308769,
                0.0017494364874437451,
                0.35424691438674927
            ]
        ]
    }
]
```

Assuming the position in the camera frame is in meters, this seems pretty ok. 

However, when I run the visualization script, I get the following result :

![all_results](https://user-images.githubusercontent.com/6552564/215747752-f9671ad0-4188-4e6f-bcb3-d16ded95b4ab.png)

Another information, the 3d reconstruction is far from perfect, but not stupidely bad. It looks like that in meshlab : 
![Capture d’écran de 2023-01-31 12-27-32](https://user-images.githubusercontent.com/6552564/215747967-2341da2e-eab0-4a3a-9876-5473408e9a4e.png)
![Capture d’écran de 2023-01-31 12-27-26](https://user-images.githubusercontent.com/6552564/215747979-8c999875-c411-4ff7-84ba-af04272410c4.png)

Any idea of what I am doing wrong ?

Thanks !

**Comment by ylabbe** [comment]
I suspect the objects may not be rendered correctly by panda3d. This can happen if the normals are incorrect (e.g. pointing in the wrong direction). Can you share the mesh ?

**Comment by apirrone** [comment]
Yes, here you go : https://drop.chapril.org/download/11f21e855935ffc8/#JL6PJRzphdj5xRUltNleYw

**Comment by apirrone** [comment]
They seem to be pointing outwards : 
![Capture d’écran de 2023-01-31 13-54-49](https://user-images.githubusercontent.com/6552564/215765627-1cd4238d-84ba-4fd2-bcd2-4fbded9d7251.png)

**Comment by ylabbe** [comment]
Great, I'm happy to see you managed to make it work!

**Comment by henry890112** [comment]
Hello apirrone, 
I meet the same problem jest like you, but I can't found how to position the cropped mesh at the origin, could you please show me how to solve it? 
Thanks!!!
![all_results](https://user-images.githubusercontent.com/105191477/220502319-ee44f4cf-5411-4e76-8dcb-f8418647ba68.png)
my following output/object_data.json:
```
[{"label": "024_bowl", "TWO": [[-0.4408988027365529, 0.07546882958549483, 0.7555218147371563, -0.4786434882846119], [-2.7438454708317295e-05, 9.068073995877057e-05, 0.00044130231253802776]]}]
```

**Comment by lcl-git-3d** [comment]
> Hello apirrone, I meet the same problem jest like you, but I can't found how to position the cropped mesh at the origin, could you please show me how to solve it? Thanks!!! ![all_results](https://user-images.githubusercontent.com/105191477/220502319-ee44f4cf-5411-4e76-8dcb-f8418647ba68.png) my following output/object_data.json:
> 
> ```
> [{"label": "024_bowl", "TWO": [[-0.4408988027365529, 0.07546882958549483, 0.7555218147371563, -0.4786434882846119], [-2.7438454708317295e-05, 9.068073995877057e-05, 0.00044130231253802776]]}]
> ```
Hello henry890112,
I meet the same problem jest like you,  could you please show me how to solve it?
Thanks!

**Comment by Windson9** [comment]
Hi @henry890112. I am also trying to run inference on custom dataset and I did also created a directory containing all the files which are required for inference. However, I am facing some issues with label IDs. Can you please share if you have done any changes in the inference script or any change in src code ?

Thanks, Mayank

**Comment by apirrone** [comment]
You can check our megapose fork at : https://github.com/pollen-robotics/megapose6d/tree/dev (dev branch)

We made a wrapper around megapose so that it is easier to use, look at `wrapper_example.py` for an example of how to use it

**Comment by Windson9** [comment]
@apirrone Thanks for sharing the code. I am curious though, can we run the inference without the wrapper which you shared ? Or I will need to train the model for that ? Also, did you train the model for your custom dataset or used the saved checkpoint provided in this repo ?

Thanks, Mayank

**Comment by lcl-git-3d** [comment]
> Hi @lcl-git-3d. I don't think the issue is with the png file. If there is a meta-data in the ply pointing to the png texture file, then only you need it. Do you get any error regarding textures when your ply file is parsed ? If not, the issue might be in your ply file. It should create a contour with no texture, either white overlay of the CAD file.

Thank you for your reply! I quite agree with you. Even if there is no texture information, there should be white areas in the visualization. I also think the error is in my ply file. My ply file was converted from the stp file using Meshlab software, so there are no errors when opening the ply file. The stp file was sent to me by my teacher and was not generated by real objects. Could you please help me check what is wrong with my ply file? Thank you very much indeed! https://drop.chapril.org/download/42dec750641f1a4b/#NcmCR97QYicH03VDlSLaAQ

**Comment by monajalal** [comment]
@apirrone I am not sure if I have same problem as you but I wonder if you could have a  look at https://github.com/megapose6d/megapose6d/issues/27 and guide me a bit?

**Comment by Windson9** [comment]
Hhi @lcl-git-3d. I will have a look at the ply file. Meanwhile, if your object of interest is the bowl. You can download the "bop_models_panda3d" from the pre-processed meshes and in that you will find the mesh for bowl in ycb dataset. If that works, you can compare both the ply files and see what's wrong. Same goes for @monajalal as well.

**Comment by lcl-git-3d** [comment]
> Hhi @lcl-git-3d. I will have a look at the ply file. Meanwhile, if your object of interest is the bowl. You can download the "bop_models_panda3d" from the pre-processed meshes and in that you will find the mesh for bowl in ycb dataset. If that works, you can compare both the ply files and see what's wrong. Same goes for @monajalal as well.

Hi,I compared my ply file with the hope_000002.ply provided by the author. The data format of these two files is the same. The difference is that after opening with Meshlab, my ply file has no texture information, but hope_000002.ply is basically the same as the real thing. Maybe there are other differences but I don't see them. I have been troubled by this problem for a long time. Could you please help me check what is wrong with my ply file? Thank you very much.

**Comment by Windson9** [comment]
Hi @lcl-git-3d please refer to this [#27](https://github.com/megapose6d/megapose6d/issues/27) issue. I think your dimensions are not in mm.

**Comment by assia855** [comment]
Hi, I follow all your suggestion but I still can't have a 3D detection. Any idea please. 
![Capture d’écran du 2023-09-06 15-17-14](https://github.com/megapose6d/megapose6d/assets/62453226/9ac248d0-2571-4530-8901-28c008cb46a1)

**Comment by BIT-TYJ** [comment]
Hello, how to generate the png of mesh please? I have one .obj mesh can be successfully used.

**Comment by monajalal** [comment]
@apirrone  were you able to get inference using camera? could you share your script?

**Comment by JJMa1907** [comment]
The issue you addressed is particularly helpful. In my specific case, maintaining precise scale (in millimeters, mm) and ensuring proper orientation are critical for achieving accurate results.
